Adds support for FMC external SDRAM preinitialisation for stm32.

k0d opened the pull request:
As a prerequisite for enabling LTDC on STM32, this pull request enables using the external SDRAM on the STM32f746g_discovery board as the main memory. It does this by configuring it at boot and then jumping to the zephyr code.

Closes #15944

Signed-off-by: Mark Olsson <mark@markolsson.se>

k0d commented:
@erwango I'll fix the failed checks on this, and then maybe you can do a code-review. There was quite a few ways to solve this issue, hope this is ok!

erwango commented on dts/arm/st/f7/stm32f7.dtsi:
```diff
@@ -652,6 +659,12 @@
 		label = "OTGFS_PHY";
 	};
+
+	sdram0: memory@c0000000 {
```
Since SDRAM is mounted on board, I'd expect this to be declared in board.dts

k0d replied:
I totally agree! Will fix that.

erwango commented on dts/arm/st/f7/stm32f7.dtsi:
```diff
@@ -639,6 +639,13 @@
 			label = "DMA_2";
 		};
+
+		fmc: stm32-fmc@60000000 {
```
@0xA0000000 ?

k0d replied:
Yes, that looks correct, I misunderstood the values in the data sheet.

erwango commented on dts/bindings/memory-controllers/st,stm32-fmc.yaml:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2020, Mark Olsson <mark@markolsson.se>
+# SPDX-License-Identifier: Apache-2.0
+
+description: STM32 Fast Memory Controller (FMC)
```
Flexible Memory...

k0d replied:
Correct :)

erwango commented on include/linker/sections.h:
```diff
@@ -66,6 +66,7 @@
 #define IMX_BOOT_DATA	.boot_hdr.data
 #define IMX_BOOT_IVT	.boot_hdr.ivt
 #define IMX_BOOT_DCD	.boot_hdr.dcd_data
+#define STM32_PREINIT	.preinit
```
If you need a change to this file, you need a specific commit and explain clearly why this is needed.
Also, it has to be generic, you can't put vendor specific code here.

k0d replied:
I can understand it should be generic, that thing is the same problem was solved for the IMX boards by adding this vendor specific code in this way. I wasn't sure if I should name it with STM32...I guess not. We need a generic way of adding preinit code that sets up external memory. That's the whole reason this was done with preinit, so it could use the OCRAM (on-chip) for the stack whilst it sets up the FMC.

erwango replied:
Ok, then, can you try to make it as generic as possible, split in commits and provide more explanations in git messages?
Whole point is to help reviewers understanding your work, your choices, what it adds and how to use it.
As an example IMX code was added here: https://github.com/zephyrproject-rtos/zephyr/pull/12237

erwango commented on soc/arm/st_stm32/common/Kconfig.soc:
```diff
@@ -8,3 +8,13 @@ DT_CHOSEN_Z_CCM := zephyr,ccm
 config STM32_CCM
 	def_bool $(dt_chosen_enabled,$(DT_CHOSEN_Z_CCM))
+
+config STM32_PREINIT
```
Please explain why this can't be done using already available mecanisms

k0d replied:
As mentioned in another comment, we need to do this before the zephyr __start method as the memory for the stack isn't available at that point. We could put the stack in OCRAM and use the FMC only for the HEAP, but that complicates things and I'm not sure how to do that.

erwango replied:
ok, replied in above comment

jinyongiot replied:
I  want to add FMC driver too! 
My idea: Adds FMC driver in the z_platform_init() function called at reset.S

k0d replied:
I tried z_platform_init, but because the stack isn't setup it can't jump back after it's executed the code inside of z_platform_init.

erwango reviewed (changes requested):
Thanks for providing this. This surely hasn't been an easy task to make this working!

As you mention, there are lot of ways to make this working, and now we'll need to find one that is possible to merge in zephyr tree.
I'm not a FMC expert, so I won't be able to say exactly what and how things should be done, but I hope I will be able to help you making this compatible with a future merge.
First step would be to explain what you tried to achieve and why you implemented it that way.
Then here are general guidelines:
- Split your changes in several commits, starting from the generic ones and ending with the most specific (board changes)
- Make thing as generic as possible, use or provide generic API to access peripherals (FMC in this case)

If you need help in some steps, don't hesitate to ask on project slack channel

k0d commented on boards/arm/stm32f746g_disco/stm32f746g_disco.yaml:
```diff
@@ -6,7 +6,7 @@ toolchain:
   - gnuarmemb
   - xtools
-ram: 256
+ram: 8192
```
The way it's wired up on the board means that only half of it is accessible. The reason I added it was because it was done that way on another board that I took my inspiration from.

https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/mimxrt1064_evk/mimxrt1064_evk.yaml

andrewboie replied:
this is going to break test case filtering which considers the amount of RAM available.
This just seems horribly wrong to me. Can't we put the actual amount of RAM in kilobytes here? what is preventing this?

k0d commented on boards/arm/stm32f746g_disco/preinit.c:
```diff
@@ -0,0 +1,155 @@
+
+	/* Disable write protection */
+	temp_register = FMC_Bank5_6->SDCR[0];
+	FMC_Bank5_6->SDCR[0] = (temp_register & 0xFFFFFDFF);
```
This all runs before the rest of Zephyr, so it can't really be in a zephyr driver if I understand correctly. The main reason I did it this way was because it's how the problem was solved for the IMX boards.

erwango replied:
Even if this is a preinit driver, it should be doable to implement it as a generic STM32 driver (could potentially be under `soc/arm/st_stm32` if there is no better location) so it could be used by more boards and socs. 
Also, the pin information configuration would be the same before and after init so it would be nice to share the same configuration (eventually provided by dts).

k0d replied:
I'll try to sort that out.

At the dev meeting on Thursday, can you discuss the early_init code (the first commit) so I at least know if we'll use that solution.

erwango replied:
ok, sure

k0d replied:
I've been doing some tests and investigation.

It seems like it'd be too much to create a driver in zephyr/drivers/memory_controller for example...and then call that in the early_init code. It requires too much of the zephyr system.

So I've created the early_init inside of soc/arm/st_stm32/stm32f7 instead. That works fine...the next step is to get it to read the config from the dts instead. That shouldn't be a problem. I will also check to see if stm32f4 etc need this and will move it if needed.

Does this sound ok to you?

erwango replied:
> I've been doing some tests and investigation.
> 
> It seems like it'd be too much to create a driver in zephyr/drivers/memory_controller for example...and then call that in the early_init code. It requires too much of the zephyr system.

Ok, thanks for having had a look. I don't think it is something achievable indeed (today at least).

> So I've created the early_init inside of soc/arm/st_stm32/stm32f7 instead. That works fine...the next step is to get it to read the config from the dts instead. That shouldn't be a problem. I will also check to see if stm32f4 etc need this and will move it if needed.
> 
> Does this sound ok to you?

Sure. Having this STM32 generic would be great indeed.

k0d commented on boards/arm/stm32f746g_disco/preinit.c:
```diff
@@ -0,0 +1,155 @@
+	configure_gpio_fmc('G', 8);
+	configure_gpio_fmc('G', 15);
+	configure_gpio_fmc('H', 3);
+	configure_gpio_fmc('H', 5);
```
These are setup before Zephyr...so the usual way is not available at that point I believe.

k0d committed:
linker: Enable early initialization of peripherals.

Some boards require the initialization of some peripherals such as external flash or sdram. Enable this and create the vector table and early_init_handler to run code at the very start of the boot_rom.

Signed-off-by: Mark Olsson <mark@markolsson.se>

k0d committed:
drivers: memory: stm32 Enables initialization of the external SDRAM on STM32f746g_discovery

Enables the use of the external SDRAM on the STM32f746g_discovery board as the main memory. It does this by configuring it at boot and then jumping to the zephyr code.

Signed-off-by: Mark Olsson <mark@markolsson.se>

galak commented:
DEV-REVIEW: Will discuss during March 26th meeting as @erwango wasn't able to join March 19th meeting.

erwango commented on boards/arm/stm32f746g_disco/early_init.c:
```diff
@@ -0,0 +1,156 @@
+	GPIOx_base->MODER |= 0x2 << (pin * 2);
+	GPIOx_base->OSPEEDR |= 0x2 << (pin * 2);
+	GPIOx_base->OTYPER |= 0x0 << (pin * 2);
+	GPIOx_base->PUPDR |= 0x1 << (pin * 2);
```
Using LL API would help you to make code STM32 generic.
You can find example in gpio_stm32.c

k0d replied:
When I include code from Zephyr or the HAL/LL api, it normally stops this from working as I don't think it can access variables stored in the .data section. I've tried to copy them, but as the linker sets things up for when the external ram is used, then trying to copy them to the on-chip ram for this early_init code, it's all wrong...often it's too big to fit also.

Any suggestions on how we can make that work?

erwango replied:
@ABOSTM, @FRASTM, maybe you know better how we can use/configure HAL to be used in this specific case ?

k0d replied:
One solution could be a small boot loader that configures the FMC...but I don't know if that will break other things?

erwango commented on boards/arm/stm32f746g_disco/early_init.c:
```diff
@@ -0,0 +1,156 @@
+
+	/* Disable write protection */
+	temp_register = FMC_Bank5_6->SDCR[0];
+	FMC_Bank5_6->SDCR[0] = (temp_register & 0xFFFFFDFF);
```
What about using HAL api (as it seems LL is not available) to make this code STM32 generic ?

k0d replied:
I can use HAL... :)

ABOSTM commented:
Hi @k0d ,
I don't really understand what you want to do with your external memory: put some code (or all code) to be executed ? or only put some data ?
Also I don't understand what is the link between LTDC and FMC ?
Nevertheless here is my point of view:
At reset, MCU will starts running from internal Flash (code) and internal RAM (data).
Then I can see several options:
1) Use only 1 global application, split into Internal Flash and external Memory, but all code and data necessary to reach FMC initialization must be in internal flash/RAM.
It doesn't necessarily mean FMC must be in early init:
For example all zephyr code/data can be put in internal Flash and internal RAM, and all user application is put in external memory (and eventually remaining internal memory).
In this example FMC initialization may occurs in regular zephyr init (as opposed to early int)

2) use a bootloader, at set at least FMC.
But then 
* either the application is still split between internal and external memory, thus we are close to option 1 and seems to more preferable to use option 1
* or whole application (Zephyr + user) is in external memory, but then it looks to me like a waste of internal flash.

As you can understand I would be in favor of option 1

k0d commented:
Hi Alexandre, Thanks for your reply!

We can just put user data (frame buffers for ltdc/screen) into the external ram.
So option 1 sounds the best.

I'll make a test to see if I can do it, otherwise I might need some help on getting this working.

tyrantlizard77 reviewed (changes requested):
I'm very uneasy about the fake 8MB of ram specified here.

k0d commented:
> I'm very uneasy about the fake 8MB of ram specified here.

The board has external SRAM (8mb), I thought it was that which we should put in the yaml file. However...my new changes to this pull request (not committed yet) don't have this...it's back as the original 256kb.

erwango commented:
@k0d, I'm moving this dev to v2.4.0, do you plan to continue it ?

k0d commented:
@erwango I will continue it...maybe in 2-3 weeks from now.

hjuul commented:
@k0d Are you still working on this?
I'd really like to see this work going into v2.4.0 :grimacing:

k0d commented:
I've not worked on this for a long time, but I do want (need) to get it working at some point. Maybe in a month or 2.

JonasVW commented:
Is there any new update on this topic in order to support the LCD on the STM32F7?

erwango commented:
@JonasVW FMC support was added here: https://github.com/zephyrproject-rtos/zephyr/pull/29686
This is H7 only for now, but can be updated for F7 support
